### ISSUE TITLE
SNOW-801097: Pyre type checker returns unexpected type errors with snowpark-python

### ISSUE THREAD

**Issue by nerdinand** [snowflakedb/snowpark-python#810]
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

```
Python 3.8.16 (default, Mar  8 2023, 23:07:01) 
[Clang 14.0.0 (clang-1400.0.29.202)]
```

2. What operating system and processor architecture are you using?

```
macOS-13.3.1-arm64-arm-64bit
```

3. What are the component versions in the environment (`pip freeze`)?

Redacted to the most relevant packages:
```
pyre-check==0.9.18
pyre-extensions==0.0.30
snowflake-connector-python==3.0.2
snowflake-snowpark-python==1.1.0
```

4. What did you do?

See: https://github.com/facebook/pyre-check/issues/716

We're using snowpark-python in conjunction with the [pyre type checker](https://pyre-check.org/). While doing this we came across some weird type errors, e.g. with this piece of code:

```python
import snowflake.snowpark.functions as spf

spf.min("foo")
```

We get the type error:

```
src/reproduction.py:3:8 Incompatible parameter type [6]: In call `spf.min`, for 1st positional argument, expected `ColumnOrName` but got `str`.
```

Since `ColumnOrName` is defined as 
```python
ColumnOrName = NewType("ColumnOrName", Union["snowflake.snowpark.column.Column", str])
```

we would not expect such an error.

As outlined in https://github.com/facebook/pyre-check/issues/716#issuecomment-1488352053, the code

```python
from typing import NewType, Union

ColumnOrName = NewType("ColumnOrName", Union["snowflake.snowpark.column.Column", str])

def min(e: ColumnOrName) -> None:
    pass

min("foo")
```

produces to two type errors:

```
ƛ Found 2 type errors!
src/reproduction.py:3:39 Invalid inheritance [39]: `Union[snowflake.snowpark.column.Column, str]` is not a valid parent class.
src/reproduction.py:8:4 Incompatible parameter type [6]: In call `min`, for 1st positional argument, expected `ColumnOrName` but got `str`.
```

This leads to our assumption that snowpark-python isn't using `NewType` correctly. Indeed, the same snippet without `NewType`:
```python
from typing import NewType, Union

ColumnOrName = Union["snowflake.snowpark.column.Column", str]

def min(e: ColumnOrName) -> None:
    pass

min("foo")
```

passes the type checks.

5. What did you expect to see?

This piece of code should not result in an error with the pyre type checker:

```python
import snowflake.snowpark.functions as spf

spf.min("foo")
```

**Comment by sfc-gh-mkeller** [comment]
I agree, according to the docs [NewType](https://docs.python.org/3/library/typing.html#newtype) is meant for distinctive types which this shouldn't be.
Great investigation @nerdinand , would you like to raise a PR to fix this issue?

**Comment by sfc-gh-mkeller** [comment]
Actually, this is a dupe of https://github.com/snowflakedb/snowflake-connector-python/issues/1525 and should be fixed by #799

**Comment by alexmojaki** [comment]
This can be closed now, right?
